document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java

```diff
@@ -0,0 +1,93 @@
+package org.hypertrace.core.documentstore.query;
+
+import java.util.ArrayList;
+import java.util.List;
+import lombok.Builder;
+import lombok.Singular;
+import lombok.Value;
+import org.hypertrace.core.documentstore.expression.FilteringExpression;
+import org.hypertrace.core.documentstore.expression.GroupingExpression;
+import org.hypertrace.core.documentstore.expression.SelectingExpression;
+
+/**
+ * A generic query definition that supports expressions. Note that this class is a more general
+ * version of {@link org.hypertrace.core.documentstore.Query}
+ *
+ * <p>Example: <code>
+ *     SELECT col4, SUM(col5) AS total
+ *     FROM <collection>
+ *     WHERE col1 < 7 AND col2 != col3
+ *     GROUP BY col4, col6
+ *     HAVING SUM(col5) >= 100
+ *     ORDER BY col7+col8 DESC
+ *     OFFSET 5
+ *     LIMIT 10
+ * </code> can be built as <code>
+ *     Query query = Query.builder()
+ *         .selection(IdentifierExpression.of("col4"))
+ *         .selection(
+ *             AggregateExpression.of(SUM, IdentifierExpression.of("col5")),
+ *             "total")
+ *         .whereFilter(LogicalExpression.of(
+ *             RelationalExpression.of(
+ *                 IdentifierExpression.of("col1"),
+ *                 LT,
+ *                 ConstantExpression.of(7)),
+ *             AND,
+ *             RelationalExpression.of(
+ *             IdentifierExpression.of("col2"),
+ *             NEQ,
+ *             IdentifierExpression.of("col3"))))
+ *         .aggregation(IdentifierExpression.of("col4"))
+ *         .aggregation(IdentifierExpression.of("col6"))
+ *         .havingFilter(
+ *             RelationalExpression.of(
+ *                 AggregateExpression.of(SUM, IdentifierExpression.of("col5")),
+ *                 GTE,
+ *                 ConstantExpression.of(100)))
+ *         .orderBy(
+ *             OrderBy.of(
+ *                 FunctionExpression.builder()
+ *                     .operand(IdentifierExpression.of("col7"))
+ *                     .operator(ADD)
+ *                     .operand(IdentifierExpression.of("col8")).build(),
+ *                 DESC))
+ *         .paginationDefinition(PaginationDefinition.of(5, 10))
+ *         .build();
+ *  </code>
+ */
+@Value
+@Builder
+public class Query {
+  List<Selection> selections;
```

Comment:
Can't we use SelectionExpression here similar to FilteringExpression , GroupingExpression?

Reply:
Selection is a combination of SelectionExpression and an alias. So, SelectionExpression can't be directly used here.

Reply:
If an alias is required for all selections, should we have this in the abstract class and have the impls of  SelectionExpression extend it?

Reply:
Even I thought of doing the same. However, there are 2 reasons why it is not done that way.

1. Not all subclasses of SelectingExpression will need an "alias" field, because a subclass of SelectingExpression might actually be used as a SortingExpression or vice versa. Here, inheritance is purely used for type-substitution. Introducing "alias" to SelectingExpression as a class will end up in not only unwanted memory, but, also change the way a Query is constructed (in an undesired manner).
2. Similar to SelectingExpression, SortingExpression can also be converted to an abstract class with the "SortingOrder" field. When we do so, classes like "AggregateExpression" will need multiple inheritance, which is not supported in Java.